### ISSUE TITLE
Optionality of specialization constants in OpenCL 3.0

### ISSUE THREAD

**Issue by AnastasiaStulova** [KhronosGroup/OpenCL-Docs#294]
The way specialization constants are described appears to be non-optional at least for devices that support IL. Can I suggest that in order to simplify adoption of OpenCL 3.0 this feature to be moved to the optional list at least for devices that support IL that is not SPIR-V but might even be for devices that support SPIR-V. Here are some of my concerns:

- Specialization constant concept don't exist yet in some other popular IR format i.e. LLVM. By making this feature non-optional we prevent using alternative to SPIR-V IR formats for OpenCL 3.0 in a compliant way.

- There is no actively used OpenCL kernel language that supports specialization constants, therefore, making it difficult to expose to the customers. Even for vendors supporting SPIR-V it might be a burden to implement functionality that their customers can't access in any way. Looking at optional language features i.e. generic address space,  we don't mandate implementing generic address space in SPIR-V even for vendors that support SPIR-V. Therefore we shouldn't force implementing a feature that doesn't have a representation in language at all not even in an optional format.

**Comment by kpet** [comment]
I agree it should be optional for devices that support an IL that is not SPIR-V but I think we should keep it as a mandatory feature when SPIR-V is supported.

>Even for vendors supporting SPIR-V it might be a burden to implement functionality that their customers can't access in any way.

They can, using SPIR-V ;).  One of the advantages of SPIR-V is that it gives users access to features independently of source languages. If a vendor ships support for SPIR-V specilisation constants now, then devices in the field will be able to take advantage of future language changes.

>Looking at optional language features i.e. generic address space, we don't mandate implementing generic address space in SPIR-V even for vendors that support SPIR-V

That's not my reading of the environment spec:

>An OpenCL environment that supports SPIR-V 1.0 must support SPIR-V 1.0 modules that declare the following capabilities: [...] GenericPointer [...]  or OpenCL 3.0 devices supporting the Generic
Address Space (where CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT is CL_TRUE)

**Comment by AnastasiaStulova** [comment]
> That's not my reading of the environment spec:
> 
>     An OpenCL environment that supports SPIR-V 1.0 must support SPIR-V 1.0 modules that declare the following capabilities: [...] GenericPointer [...] or OpenCL 3.0 devices supporting the Generic
>     Address Space (where CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT is CL_TRUE)

Does this mean that devices that don't support generic address space in OpenCL 3.0 but support SPIR-V have to implement it anyway? I don't even know if it will even be possible but more importantly why would OpenCL make such a requirement? How is this enforced currently i.e. would CTS check this or something? I don't know why we need to force implementing optional features in the IR if they have no use. In OpenCL 3.0 we provide special queries to gather what capabilities are supported by the device. I imagine an application developer could just check them before importing the SPIR-V or alternatively an automated approach can be implemented that would check what functionality is being used in SPIR-V module against supported features. As a matter of fact we already have such functionality implemented in SPIRV-LLVM Translator.

**Comment by alycm** [comment]
Having specialization constants be mandatory if SPIR-V is supported was an active decision rather than something that slipped in.  Having specialization constants affect other IL's may be an oversight that has been inherited from OpenCL 2.2, but I'll discuss that in a separate issue.

I'd agree with the argument that SPIR-V is the source language in question from an OpenCL perspective, and colleagues have done experiments with specialization constants in the on top of OpenCL without touching OpenCL C: https://github.com/codeplaysoftware/standards-proposals/blob/master/spec-constant/index.md

A specialization constant concept could be added to OpenCL C, it doesn't help much for JIT compiled code, but could be interesting for compiling OpenCL C to SPIR-V.

For the generic address space comparison I think the 3.0 provisional SPIR-V environment spec is clear that the `GenericPointer` capability only needs to be supported for devices that report `CL_​DEVICE_​GENERIC_​ADDRESS_​SPACE_​SUPPORT` as `CL_​TRUE`.   It is the part that @kpet quoted, but we seem to be taking an opposite reading.

**Comment by alycm** [comment]
Non-SPIR-V IL point is in issue https://github.com/KhronosGroup/OpenCL-Docs/issues/303, so this issue can focus on whether the optionality should be independent of IL support.

**Comment by AnastasiaStulova** [comment]
> I'd agree with the argument that SPIR-V is the source language in question from an OpenCL perspective, and colleagues have done experiments with specialization constants in the on top of OpenCL without touching OpenCL C: https://github.com/codeplaysoftware/standards-proposals/blob/master/spec-constant/index.md

For vendors that want only to implement OpenCL 3.0 how would SPIR-V specialization constants be used? Do they then need to make sure other languages compile to their targets? I would see that the support of other languages should be an optional feature?

**Comment by bashbaug** [comment]
Hi @AnastasiaStulova , can you please say more what you mean by:

> For vendors that want only to implement OpenCL 3.0 how would SPIR-V specialization constants be used?

The way I see it there are only two OpenCL 3.0 scenarios that are relevant to specialization constants:

1. For OpenCL 3.0 implementations that support SPIR-V IR, these implementations currently must support specialization constants.  There is currently no way to express specialization constants in OpenCL C, but SPIR-V specialization constants can be generated via other mechanisms.

2. For OpenCL 3.0 implementations that do not support SPIR-V IR, support for specialization constants is not required, as described in the [appendix](https://bashbaug.github.io/OpenCL-Docs/html/OpenCL_API.html#_intermediate_language_programs).

Is there another scenario to consider also?  Thanks!

**Comment by AnastasiaStulova** [comment]
@bashbaug Regarding **1**, I don't imagine we expect the application developers to write SPIR-V assembly with spec constants. Therefore I don't understand why would we require to implement spec constants for which there is no active OpenCL kernel language. It seems to be a better fit as an optional feature. I understand that some non-OpenCL languages could allow the use of spec constants, but non-OpenCL kernel languages are an optional feature? It seems inconsistent to me that some features i.e. generic address space that we can be represented in a kernel language are not mandated to be implemented in SPIR-V but spec constants that have no representation in OpenCL C 3.0 are.

Regarding **2** my current interpretation is that we require spec constants for any IL, but I think this is being discussed in https://github.com/KhronosGroup/OpenCL-Docs/issues/303 now.

**Comment by StuartDBrady** [comment]
Note that for the **Kernel** execution model, the SPIR-V specification describes **OpSpecConstantOp** operations using pointer/address operands, the result of which could then be used for other purposes e.g. in determining array lengths in **OpTypeArray**.

Consider the following sequence of instructions:
* **OpSpecConstantOp OpConvertPtrToU** used to convert a pointer to a global to an integer.
* **OpBitwiseAnd** used to extract the least significant bit of the pointer, to determine its alignment.
* **OpTypeArray** used to declare an array type with a number of elements determined by the pointer's alignment.

In my opinion, it is in no way reasonable to expect that the addresses held in pointers will be known (or knowable) at the time of specialization.

I would note the SPIR-V specification also describes floating-point operations for **OpSpecConstantOp** when using the **Kernel** execution model. It is typical to implement compile-time evaluation of floating-point operations using a software library, rather than relying on the floating-point unit of the system running the compiler (or in this case, the specializer). Otherwise, the precise behavior of the compiler (or specializer/translator) will be dependent on the system upon which it runs, with respect to rounding and the handling of NaNs. Determinism of the compiler (or specilizer/translator) is a desirable quality, and as such, implementing **OpSpecConstantOp** for floating-point operations to a reasonable standard would require adding extra dependencies to the _Khronos SPIR-V/LLVM Translator_, which is not a decision that should be made lightly.

On this basis, I must ask that an exemption be made in the _OpenCL SPIR-V Environment Specification, such that the pointer-related operations need not be implemented for **OpSpecConstantOp**. I would also _request_ the same for the floating-point operations to lessen the burden for implementations currently targeting OpenCL 2.1, looking to offer OpenCL 3.0 support with SPIR-V, if there are no strong objections.

**Comment by bashbaug** [comment]
We discussed this issue in the June 30th teleconference.  In short, we decided that if an OpenCL 3.0 implementation supports SPIR-V then it will also support specialization constants.

We will create a separate issue to address the cases described in @StuartDBrady's [comment above](https://github.com/KhronosGroup/OpenCL-Docs/issues/294#issuecomment-648970076).  Once this issue is created, we will close this issue with no additional action, since the spec already requires specialization constants when the implementation supports SPIR-V.
